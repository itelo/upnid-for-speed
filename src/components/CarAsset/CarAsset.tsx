import React from "react";
import { motion } from "framer-motion";
import styled from "styled-components";
import carImg from "../../assets/car.png";

type CarAssetProps = {
  size: number;
  matrix: { x: number; y: number; scale: number }[][];
  position: [number, number];
} & React.ImgHTMLAttributes<HTMLImageElement>;

const CarAssetStyled = styled(motion.img)<{
  size: number;
}>`
  top: ${({ size }) => (size * 70)/100}px;
  left: ${({ size }) => (size - (size * 28)/100) / 2 - ((size * 2)/100)}px;
  height: ${({ size }) => (size * 28)/100}px;
  width: ${({ size }) => (size * 28)/100}px;
  position: absolute;
`;

const CarAsset = (props: CarAssetProps) => {
  return (
    <CarAssetStyled
      data-testid="car-asset"
      alt="car"
      size={props.size}
      animate={{
        translateX:
          props.matrix[props.position[1]][props.position[0]].x * props.size,
        translateY:
          props.matrix[props.position[1]][props.position[0]].y * props.size,
        scale: props.matrix[props.position[1]][props.position[0]].scale
      }}
      src={carImg}
    />
  );
};

CarAsset.defaultProps = {};

export default CarAsset;
