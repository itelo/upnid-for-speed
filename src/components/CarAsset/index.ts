export {default as CarAsset } from './CarAsset';