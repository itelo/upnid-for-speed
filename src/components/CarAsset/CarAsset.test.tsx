import React from "react";
import { render, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import CarAsset from "./CarAsset";
import "jest-styled-components";

describe("<CarAsset />", () => {
  const matrix = [
    [
      { x: -1, y: -1, scale: -1 },
      { x: 0.5, y: 0.5, scale: 0.5 },
      { x: 1.2, y: 1.2, scale: 1.2 }
    ]
  ] as { x: number; y: number; scale: number }[][];
  const leftPosition = [0, 0] as [number, number];
  const middlePosition = [1, 0] as [number, number];
  const rightPosition = [2, 0] as [number, number];

  it("test basic render at middle", async () => {
    const { getByTestId } = render(
      <CarAsset size={100} matrix={matrix} position={middlePosition} />
    );

    const element = getByTestId("car-asset");

    await waitFor(() =>
      expect(element).toHaveStyle("transform: translateX(50px) translateY(50px) scale(0.5) translateZ(0);")
    );
  });

  it("test basic render at left", async () => {
    const { getByTestId } = render(
      <CarAsset size={100} matrix={matrix} position={leftPosition} />
    );

    const element = getByTestId("car-asset");

    await waitFor(() =>
      expect(element).toHaveStyle("transform: translateX(-100px) translateY(-100px) scale(-1) translateZ(0);")
    );
  });

  it("test basic render at right", async () => {
    const { getByTestId } = render(
      <CarAsset size={100} matrix={matrix} position={rightPosition} />
    );

    const element = getByTestId("car-asset");

    await waitFor(() =>
      expect(element).toHaveStyle("transform: translateX(120px) translateY(120px) scale(1.2) translateZ(0);")
    );
  });

  it("test snapshot", () => {
    const { container } = render(
      <CarAsset size={100} matrix={matrix} position={middlePosition} />
    );

    expect(container.firstChild).toMatchSnapshot()
  })
});