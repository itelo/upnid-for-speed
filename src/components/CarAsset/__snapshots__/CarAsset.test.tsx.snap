// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CarAsset /> test snapshot 1`] = `
.c0 {
  top: 70px;
  left: 34px;
  height: 28px;
  width: 28px;
  position: absolute;
}

<img
  alt="car"
  class="c0"
  data-testid="car-asset"
  size="100"
  src="car.png"
/>
`;
